### ISSUE TITLE
Disabling Rollback and Custom S3 Bucket/Key

### ISSUE THREAD

**Issue by Spareo** [cloudtools/stacker#631]
I just have a couple questions for some things I haven't been able to find. I have dug around the source a bit but just wanted to verify here.

1. Is it possible to disable the rollback feature if the stack fails to deploy? We have a lot of stuff that goes on in UserData and if that fails, we like to get on the EC2 instance and troubleshoot. With rollback, it kills the instances so we can't do this anymore.

2. I am aware that you can set a custom bucket for where the template is uploaded, but is it possible to also set the Key so it goes into specific folders in the bucket?

**Comment by xiaket** [comment]
To answer your question:

1. I think the current codebase does not support this feature. To disable rollback, we must specify `OnFailure='DO_NOTHING'` during stack creation. When we build the arguments for Cloudformation in `generate_cloudformation_args` in `~/providers/aws/default.py` file, we have not included this flag at all. Correct me if I'm wrong, but I think you cannot disable rollback for an existing stack.

I think it is possible to add support for this feature, reading environment variables would be a quick fix, but I don't know whether that's a good idea. I think it's more about design than the implementation.

2. I think the path is hardcoded in `stack_template_key_name` in `actions/base.py` file. If you are familiar with Python, I guess you could somehow override this function with a wrapper around stacker command. Again, this could be customisable, but I don't think it's a good idea. Mixing other things with cloudformation templates in a single bucket does not make sense to me, but that's just my personal opinion.

**Comment by Spareo** [comment]
You're right you cant disable rollback for an existing stack, it has to be passed in when creating the stack. I was thinking it could be a property on the stack objects in the configuration file like the enabled property. True by default but set false to disable rollback.

**Comment by xiaket** [comment]
@Spareo Excuse me but why was this issue closed? I guess we don't have a solution yet?

**Comment by Spareo** [comment]
I assumed it was dead since I never got anymore replies and just accepted that I can’t set the bucket key.

**Comment by Spareo** [comment]
Ive since just accepted the fact that we can't provide a key and worked with it. So its not a huge priority for us either. 

Disabling the rollback would be a bigger deal for my team. We work with a legacy platform that is still pegged to windows. The biggest use case for us here is when we have EC2 instances in the stack and the stack fails, we will remote into the instance and look over logs, in particular the ec2 config log since the failure usually happens while the Userdata script is executing. We will also adjust the script on the instance and retry it so we don't have to redeploy each time to test minor changes. 

The other situation where not rolling back has been handy is troubleshooting the deployment of Windows Server 2016 with docker images and getting the ec2 configuration and deployment scripts correct. 

Normally we just log into the AWS console and kill the stack after we are done reviewing it. By the way, all these scenarios are only played out in our dev environment. The other way we can kill stacks is through our gitlab runners which allow you to stand up and bring down environments. You can execute whatever scripts you want when you bring down the environment, including killing the stacks.
